nifi-nar-bundles/nifi-scripting-bundle/nifi-scripting-processors/src/main/java/org/apache/nifi/processors/script/ScriptedPartitionRecord.java

```diff
@@ -0,0 +1,232 @@
+            final ScriptEvaluator evaluator
+    ) {
+        final RecordReaderFactory readerFactory = context.getProperty(RECORD_READER).asControllerService(RecordReaderFactory.class);
+        final RecordSetWriterFactory writerFactory = context.getProperty(RECORD_WRITER).asControllerService(RecordSetWriterFactory.class);
+        final Map<String, String> originalAttributes = incomingFlowFile.getAttributes();
+
+        try {
+            session.read(incomingFlowFile, new InputStreamCallback() {
+                @Override
+                public void process(final InputStream in) throws IOException {
+                    try (
+                        final RecordReader reader = readerFactory.createRecordReader(originalAttributes, in, incomingFlowFile.getSize(), getLogger())
+                    ) {
+                        final RecordSchema schema = writerFactory.getSchema(originalAttributes, reader.getSchema());
+                        final RecordSet recordSet = reader.createRecordSet();
+                        final PushBackRecordSet pushBackSet = new PushBackRecordSet(recordSet);
+                        final Map<String, RecordBatchingProcessorFlowFileBuilder> recordSetFlowFileBuilders = new HashMap<>();
+                        final BiFunction<FlowFile, OutputStream, RecordSetWriter> recordSetWriterFactory = (outgoingFlowFile, out) -> {
+                            try {
+                                return writerFactory.createWriter(getLogger(), schema, out, outgoingFlowFile);
+                            } catch (final IOException | SchemaNotFoundException e) {
+                                throw new ProcessException("Could not create RecordSetWriter", e);
+                            }
+                        };
+
+                        int index = 0;
+
+                        // Reading in records and evaluate script
+                        while (pushBackSet.isAnotherRecord()) {
+                            final Record record = pushBackSet.next();
+                            final Object evaluatedValue = evaluator.evaluate(record, index++);
+                            getLogger().debug("Evaluated scripted against {} (index {}), producing result of {}", record, index - 1, evaluatedValue);
+
+                            if (evaluatedValue != null && evaluatedValue instanceof String) {
+                                final String partition = (String) evaluatedValue;
+
+                                if (!recordSetFlowFileBuilders.containsKey(partition)) {
+                                    recordSetFlowFileBuilders.put(partition, new RecordBatchingProcessorFlowFileBuilder(incomingFlowFile, session, recordSetWriterFactory));
+                                }
+
+                                final int recordCount = recordSetFlowFileBuilders.get(partition).addRecord(record);
+                                session.adjustCounter("Record Processed", recordCount, false);
```

Comment:
I think this logic is incorrect. If you have 3 records, then for the first record you'll adjust the counter by 1. The second time you'll adjust the counter by 2, and the third time you'll adjust it by 3. So you'll have called adjustCounter for 1 + 2 + 3 = 6 instead of adjustCounter(3).
Should either use `session.adjustCounter("Records Processed", 1, false);` or wait until the end and call `session.adjustCounter("Records Processed", writeResult.getRecordCount(), false);` I would recommend the latter, as it keeps the logic cleaner - do the processing, and at the end handle updating the counters.

Would also make sure that the name is Records with an s.

Reply:
You are perfectly right, I missed that this is an incrementing method and not a setting one.